FR #962: Adding Correction Handling attribute (#1368)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/correction_handling.md

Reviewer: Discussion in Sep 9 TF1: we need to be more clear what "immutable" means.

Author: Immutability is mentioned in several places within the Correction Handling and Invoice Handling attributes. It is currently tied to issued invoices and finalized charge records.

I think we should maintain immutability for issued invoices, but for finalized charge records we have a couple of options:

* Distinguish invoice-relevant vs non-invoice-relevant dimensions/metrics:
  * Apply stronger immutability restrictions to elements that appear on the invoice.
  * The challenge is identifying exactly which dimensions and metrics are present on the invoice.
* Tie immutability only to the billing period and billed cost:
  * In this approach, the content of the invoice and the dataset artifact may differ.

Consequently, stronger restrictions would apply to immutable elements within charge records, while for other elements, we could suggest that a Replacement approach is allowed if the user chooses.

Still not sure how to account for technical issues (see the example @rileyjenk mentioned in [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368#discussion_r2310555964) )

Reviewer: Discussed in Sep 16 TF1.  @ijurica will start with a list of proposed immutable columns.  We would at a minimum start with:

- Invoice ID
- Billing Period Start
- Billing Period End
- Billed Cost

Immutability here would be that they must match what's listed on the invoice -- and there may be some technical problem with the FOCUS dataset that may need to be fixed, which would require it being mutable _in order to get the dataset and the invoice to match_.

Author: We no longer use the term immutable, even in the context of issued invoice descriptions. For example, rather than saying that "*...an issued invoice is considered immutable...*", the current version includes the following Invoice Handling normative requirements, relevant for this conversation:

> * *Invoice reconciliation* MUST include (but is not limited to) the following metrics and dimensions: BilledCost, BillingCurrency, InvoiceId, InvoiceIssuer, BillingAccountId, BillingPeriodStart, and BillingPeriodEnd.
> * *Invoice reconciliation* MUST be performed by the Invoice Issuer before issuing an invoice.
> * Cost and usage data presented on an invoice and included in *invoice reconciliation* MUST be documented by the Invoice Issuer and accessible to practitioners.
> * Cost and usage data presented on an *issued invoice* MUST NOT be altered.

Regarding charge records, modifications (updates, additions, or omissions) are allowed, but under specific conditions:

> * Correction to *issued charges* adheres to the following additional requirements:
>   * Correction MUST NOT be applied when it results in discrepancies with the cost and usage data presented on the associated *issued invoice*.
>   * Correction SHOULD NOT be applied when it does not results in discrepancies with the cost and usage data presented on the associated *issued invoice*, but still affects downstream FinOps capabilities (e.g., chargeback).
>   * Correction MAY include additional charges associated with a different InvoiceId than the original charges.
>   * Correction MAY include additional charges associated with a different billing period than the original charges.
> * Correction to previously *closed billing period* that requires issuing additional invoices MUST result in additional charge(s) associated with a subsequent *open billing period*, with the charge period indicating when the cost was incurred.
> 

---

*Note: I would primarily like to get feedback on the proposed concept. Of course, I’m open to suggestions on how to rephrase.*

Corrections may arise from a variety of operational or technical causes, such as refunds, delayed or missing cost and usage data, rounding errors, post-processing adjustments, etc.

### Business Requirements and Constraints in Invoiced Billing Periods

Accurate correction handling is essential for a range of business-critical processes, including but not limited to:

* Temporal accuracy - capturing both when the charge was incurred (reflected in charge period columns, i.e., Charge Period Start and Charge Period End) and when the correction was invoiced (reflected in billing period columns, i.e., Billing Period Start and Billing Period End columns).

> * Financial and legal integrity - preserving original charge records associated with finalized invoices while recording changes to those records separately, as finalized invoices represent binding financial documents requiring immutability and traceability.

Reviewer 2: I am not sure how strong a word immutability is. It is a valid scenario to completely cancel an issued invoice and re-issue a new invoice without replacing cost and usage records. As long as that scenario is considered immutable - we are good

Reviewer: Mentioning / incorporating exceptions around immutability is important.

Reviewer: Discussed in Sep 16 TF1.  Will close in favor of the file-level comment talking about how to resolve immutability.

* Cost allocation and chargeback - attributing corrections to the correct dimensions (e.g., Billing Account, Sub Account, SKU ID, SKU Price ID, Resource ID).
* Auditability - tracing the full lifecycle of a charge from the original record through all subsequent corrections.

Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associated with an invoice once it has been issued. This ensures that issued invoices and their underlying charge records remain immutable for financial, auditing, and compliance purposes.

> A billing period is considered invoiced (or closed) once all invoices for that period have been issued and all charge records for that period are finalized. After a billing period is invoiced, no new charge records may be associated with it, and all previously finalized charge records remain unchanged. Any necessary corrections to charges originally incurred in an invoiced billing period must instead be reflected in a subsequent open billing period, with the charge period indicating when the cost was incurred. This provides a clear temporal boundary between billing cycles, preserving immutability while still allowing corrections to be tracked transparently in later billing periods.

Reviewer 2: There might be more to this. For a scenario, where the billing period is invoiced but the customer has not paid for it yet. Sometimes that customers might prefer to restate the invoice for the old billing period itself. Not replace but restate with a new invoice id. I believe this statement would not allow for that flexibility - since it says "must instead be reflected in a subsequent open billing period"

Author: The exception had already been added to the Exception section, and in the current version it’s also included in this intro section 

> Exceptionally, additional invoices may be issued for a closed billing period only if explicitly requested by the end-user.

I hope this covers the scenario described here, so I’m closing this conversation. That said, if you feel the concern hasn’t been fully addressed, please post new feedback on the latest version or feel free to reopen this thread.

* Accounting-style correction generally follows a two-step representation. Depending on the nature of the correction, either or both of the following steps may be required: (1) reversal of the original record using a charge in which cost- and quantity-related columns carry values with the opposite sign, while all other columns match the original; and (2) a new record with corrected values. This style preserves full correction history.

### Data Integrity

To ensure data integrity, correction records must not result in double counting of any cost- or quantity-related values. This applies regardless of the correction style or delivery mechanism used.

## Attribute ID

CorrectionHandling

## Attribute Name

Correction Handling

## Description

Defines how updates to previously provided [*charges*](#glossary:charge) are represented in FOCUS datasets.

## Requirements

All corrections adhere to the following requirements:

> ### Invoice and Billing Period Requirements

Reviewer: Discussed in Sep 2 TF1.  @ijurica will move this section into a separate `Invoice Handling` markdown -- either here in this PR, or in the parallel-path PR #1082.

Author: Invoice- and billing period-related normative requirements have been relocated from Correction Handling to Invoice Handling, and the headers in the Correction Handling Requirements section have been removed.

---
resolving this conversation

> * Invoice MUST be considered finalized and immutable once issued.

Reviewer 3: How can we have an invoice requirement in a cost dataset? That needs to be a requirement on the invoice itself. We can't enforce that here. Unless this moves to an "Assumptions" section.

Author: resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

* Once the associated invoice is issued, each underlying *charges* adheres to the following additional requirements:

>   * *Charge* MUST be considered finalized and immutable.

Reviewer 3: This also feels like an assumption. (Although the next one is definitely a requirement on the dataset.)

Author: resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

  * *Charge* MUST NOT be updated, deleted, or omitted.
* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
* Additional *charges* MUST NOT be associated with an invoice once it is issued.

> * Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.

Reviewer 3: nit: Should these be requirements on the InvoiceId and BillingPeriod* columns? I'm on the fence whether these are assumptions or hard requirements we can hold data generators accountable for.

Reviewer: There are certainly some requirements here that belong in an Invoice Handling attribute, which is forthcoming per @Matt-Cowsert and PR #1321.

Hopefully we can get that PR merged this week, and then @ijurica can propose changes.  Otherwise, she'll need to stand up a separate Invoice Handling here in this PR, and then we'll have to reconcile later.

Author: resolving in favor of [this conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368/files#r2352274007)

> ### General Requirements for Corrections

Reviewer: Discussed in Sep 2 TF1.  @ijurica will strike the headers in the Requirements section.

Author: Invoice- and billing period-related normative requirements have been relocated from Correction Handling to Invoice Handling, and the headers in the Correction Handling Requirements section have been removed.

---
resolving this conversation

> * Additional *charges* MUST NOT be associated with a billing period once it is invoiced and closed.

Reviewer 2: same as my comment above. Sometimes customers prefer the changes done to existing billing period if the payment was done yet

Author: As I mentioned earlier in [this ](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1368#discussion_r2399569914)thread, in the Invoice Handling - Exception section we specify:

> * Exceptions to the restrictions on *issued invoices* and *closed billing period* MAY apply in the following cases:
>   * Upon explicit request from the end-user (subject to validation and approval processes).

Regarding the individual normative requirements, although for a while were mentioning the exception directly in relevant requirements, after several discussions with TF-1 members and maintainers, we landed on keeping it in the Exception section without explicitly mentioning it in a particular requirement - at least for now (the same approach applied in other attribute definitions).

With that, I’m closing this conversation. Of course, please feel free to reopen it if you feel this needs further discussion.

* Correction handling implementation MUST support auditability by enabling traceability from the original record through all subsequent corrections for invoiced billing periods.
* Correction handling implementation MUST ensure that the delivery mechanisms and correction styles in use are documented within the Data Generator documentation.
* ChargeClass MUST be null when the *charge* does not represent a correction to a previously invoiced *billing period*.
* Correction MUST NOT result in double counting of any cost- or quantity-related values.

### Constraints in an Invoiced Billing Period

> * Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:
>   * ChargeClass MUST be "Correction".
>   * Correction MUST NOT replace or omit the original record.
>   * Corrected row(s) MUST be assigned to a different InvoiceId than the original record.
>   * [BillingPeriodStart](#billingperiodstart) and [BillingPeriodEnd](#billingperiodend) MUST equal the [*inclusive start bound*](#glossary:inclusivestartbound) and [*exclusive end bound*](#glossary:exclusiveendbound) of a subsequent open billing period in which the correction is issued.
>   * [ChargePeriodStart](#chargeperiodstart) and [ChargePeriodEnd](#chargeperiodend) MUST equal the *inclusive start bound* and *exclusive end bound* of the period in which the cost was originally incurred.

Reviewer 3: ```suggestion
Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:

* ChargeClass MUST be "Correction".
* Correction MUST NOT replace or omit the original record.
* Corrected row(s) MUST be assigned to a different InvoiceId than the original record.
* [BillingPeriodStart](#billingperiodstart) and [BillingPeriodEnd](#billingperiodend) MUST equal the [*inclusive start bound*](#glossary:inclusivestartbound) and [*exclusive end bound*](#glossary:exclusiveendbound) of a subsequent open billing period in which the correction is issued.
* [ChargePeriodStart](#chargeperiodstart) and [ChargePeriodEnd](#chargeperiodend) MUST equal the *inclusive start bound* and *exclusive end bound* of the period in which the cost was originally incurred.
```

Author: Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

### Constraints in the Replacement Mechanism

> * Replacement mechanism adheres to the following additional requirements:
>   * Correction handling implementation SHOULD support optional external retention of historical snapshots to enable traceability in uninvoiced billing periods (including both previous and current).
>   * Corrections for previously invoiced billing periods MUST be represented exclusively through the addition of new records.
>   * Corrections for uninvoiced billing periods MAY include updates, additions, or omissions.

Reviewer 3: ```suggestion
Replacement mechanism adheres to the following additional requirements:

* Correction handling implementation SHOULD support optional external retention of historical snapshots to enable traceability in uninvoiced billing periods (including both previous and current).
* Corrections for previously invoiced billing periods MUST be represented exclusively through the addition of new records.
* Corrections for uninvoiced billing periods MAY include updates, additions, or omissions.
```

Author: Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

### Constraints in the Append-only Mechanism

> * Append-only mechanism adheres to the following additional requirements:
>   * All previously delivered records MUST be retained without modification or deletion.
>   * All corrections MUST be represented exclusively by adding new records.

Reviewer 3: ```suggestion
Append-only mechanism adheres to the following additional requirements:

* All previously delivered records MUST be retained without modification or deletion.
* All corrections MUST be represented exclusively by adding new records.
```

Author: Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

## Exceptions

> * Exceptions to the restrictions on issued invoices, invoiced billing periods, and finalized charge records MAY apply in the following cases:
>   * Upon explicit request from the end-user (subject to validation and approval processes).
>   * Due to technical issues encountered during or after invoice issuance or billing period closure.

Reviewer 3: ```suggestion
Exceptions to the restrictions on issued invoices, invoiced billing periods, and finalized charge records MAY apply in the following cases:

* Upon explicit request from the end-user (subject to validation and approval processes).
* Due to technical issues encountered during or after invoice issuance or billing period closure.
```

Author: Although the wording is not optimal, the parent bullet in the original text inherently carries a condition that applies to all nested bullets.

While we don’t yet have formal guidelines for Attribute normative requirements, in the Column normative requirements guidelines we explicitly outlined that a parent bullet carries a condition applicable to its nested bullets. I’d prefer to keep the same approach here.

---

@flanakin I’m keeping the original format (for now) and resolving the conversation. However, if you feel it’s important to adjust the format, feel free to unresolve.

If a correction is applied to a charge from a closed billing period, the corrected record must:

* The correction MUST NOT replace or omit the original record.
* Be assigned to a different invoice (i.e., not the one originally associated with the charge).
* Have BillingPeriodStart and BillingPeriodEnd values that correspond to the open billing period in which the correction is issued.
* Preserve the original ChargePeriodStart and ChargePeriodEnd values to indicate when the corrected cost was actually incurred.

This dual treatment—distinguishing the charge period from the billing period—allows cost data consumers to understand both the historical intent and the current accounting context for each correction.

## Attribute ID

CorrectionHandling

## Attribute Name

Correction Handling

## Description

Defines how updates to previously provided charge records are represented in FOCUS datasets.

## Requirements

> > WORK IN PROGRESS !!!

Reviewer: Need to scrub this line if this PR is truly out of draft.

All corrections/changes to previously provided charge records in FOCUS dataset MUST follow the correction handling requirements listed below.

* Corrections to charges in previously closed billing periods MUST satisfy the following:
  * The correction MUST NOT replace or omit the original record.
  * The corrected row(s) MUST be assigned to a different InvoiceId than the one associated with the original record.
  * The BillingPeriodStart and BillingPeriodEnd MUST reflect the current open billing period in which the correction is being issued.
  * The ChargePeriodStart and ChargePeriodEnd MUST retain the original time interval in which the charge occurred.
  * ChargeClass MUST be set to "Correction".
* Providers MUST clearly document which provisioning and correction styles are in use (Replacement, Ledger-style, Accounting-style).
* Etc.

## Exceptions

> > WORK IN PROGRESS !!!

Reviewer: Need to scrub this line if this PR is truly out of draft.

> # Correction Handling

Reviewer: I think that this write-up is long enough that you need to introduce some subheaders for ease of reading.

Author: addressed. resolving.

Corrections may arise from a variety of operational or technical causes, such as refunds, delayed or missing cost and usage data, rounding errors, post-processing adjustments, etc.

Accurate correction handling is essential for a range of business-critical processes, including but not limited to:

* Temporal accuracy - capturing both when the charge was incurred (reflected in charge period columns, i.e., Charge Period Start and Charge Period End) and when the correction was invoiced (reflected in billing period columns, i.e., Billing Period Start and Billing Period End columns).
* Financial and legal integrity - preserving original charge records associated with finalized invoices while recording changes to those records separately, as finalized invoices represent binding financial documents requiring immutability and traceability.
* Cost allocation and chargeback - attributing corrections to the correct dimensions (e.g., Billing Account, Sub Account, SKU ID, SKU Price ID, Resource ID).
* Auditability - tracing the full lifecycle of a charge from the original record through all subsequent corrections.

> Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associated with an invoice once it has been issued. This ensures that issued invoices and their underlying charge records remain immutable for financial, auditing, and compliance purposes.

Reviewer 4: This is speaking in terms of the invoice being unchanged, which is different that the Cost and Usage Dataset being unchanged.

Author: I’d like to discuss the so-called Corrections to Cost vs Corrections to Usage topic that @rileyjenk mentioned last week outside the TF-1 meeting. As I understand it, this is about data corrections that don’t impact invoices. 

It would be great if @rileyjenk  could elaborate a bit more on that...

Reviewer: Discussed in Sep 2 TF1.  @rileyjenk will provide @ijurica some examples that do not fit the framing as laid out here; Irena will then revise the verbiage to handle.  Irena requests that these examples include how to denote Billing Period Start and End.

Author: See scenarios S-1 and S-1-B in this [sheet](https://docs.google.com/spreadsheets/d/1UDZCxPqUNEUQt90h8sW-YuhgBsk4pHYcwRlgPJVmwPo/edit?gid=1769206673#gid=1769206673).

Author: These paragraphs have been reworded - attaching the current version below:

> Once an invoice is issued, it becomes the authoritative financial document. Issued invoice is considered finalized, and the financial data it contains must not be altered. While modifications to the underlying cost and usage charges associated with an issued invoice (e.g., updates, additions, or omissions) may be permitted, they must not compromise the integrity of the issued invoice. Only modifications that maintain alignment with the invoice content are acceptable. Any misalignment would invalidate the prior *invoice reconciliation* and undermine the invoice's financial validity.
> 
> Corrections to the underlying cost and usage charges associated with an issued invoice that do not impact data presented on the invoice are allowed. However, although these modifications do not affect *invoice reconciliation*, they can still result in loss of auditability and traceability, which in turn complicates modifications and mappings required in downstream FinOps activities, such as cost allocation, chargeback, or budgeting. For this reason, such modifications are not preferred and should only be applied when explicitly requested by the end-user.

Also, [here](https://docs.google.com/spreadsheets/d/1UDZCxPqUNEUQt90h8sW-YuhgBsk4pHYcwRlgPJVmwPo/edit?gid=1769206673#gid=1769206673) you can find both the scenario and the sample data that were used as the basis for the discussion.

I’m closing this conversation, but if you feel the concern still hasn’t been properly addressed, please post suggestion or feedback on the latest version or feel free to reopen this thread.

> Once an invoice is issued, it serves as the authoritative financial document and is considered finalized and immutable. All charge records associated with an issued invoice are also considered finalized and must remain unchanged (i.e., corrections to finalized charge records, whether as updates, deletions or omissions, are not permitted). Furthermore, no additional charge records may be associated with an invoice once it has been issued. This ensures that issued invoices and their underlying charge records remain immutable for financial, auditing, and compliance purposes.

Reviewer 4: This does not work. What if the invoice issuer has $100 on it but the Cost and usage data only has $90 because an entry was mistakenly not populated in the data. In order to fix it I need the ability to add a row that was excluded. There was no financial correction in a followup period. I got invoiced, it just wasn't in the Cost and Usage dataset.

Author: I'd argue this falls under: Exceptions … due to technical issues encountered during or after invoice issuance or billing period closure.

I agree that that exception is rather vague. I could try to rephrase it to make it more specific, but I’m hesitant in case I leave something out. 😞

A billing period is considered invoiced (or closed) once all invoices for that period have been issued and all charge records for that period are finalized. After a billing period is invoiced, no new charge records may be associated with it, and all previously finalized charge records remain unchanged. Any necessary corrections to charges originally incurred in an invoiced billing period must instead be reflected in a subsequent open billing period, with the charge period indicating when the cost was incurred. This provides a clear temporal boundary between billing cycles, preserving immutability while still allowing corrections to be tracked transparently in later billing periods.

> FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only.

Reviewer: I suggest we bold these terms.
```suggestion
FOCUS supports two cost and usage data delivery mechanisms: **Replacement** and **Append-only**.
```

Author: I’ve added subheaders. It seems to me that bolding is no longer critical (since we haven’t used bolding anywhere else). 
@shawnalpay If you feel we should still keep bolding, feel free to reopen the conversation.

> FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only.

Reviewer 4: There is an assumption here that isn't accurate. For many billing systems there exists the presence of a "we update the data to meet our customer needs" and that can take many forms. For example a vendor may do a partition update where their typical update pattern is to append additional partitions and update existing partitions.

Author: I believe the described scenario falls into Replacement mechanism, which doesn't exclude appending.

Reviewer: Discussed in Sep 2 TF2.  @ijurica to mention this scenario under `Replacement`.

Author: The introductory section has been updated. I hope this now covers the hybrid scenario described above. 

> FOCUS supports two cost and usage data delivery mechanisms: Replacement and Append-only. **These mechanisms are not mutually exclusive and hybrid implementations are common, allowing Data Generators to meet specific delivery and auditability requirements.**

---

Resolving the conversation for now. Feel free to reopen and provide suggestions if additional adjustments are needed.

---

Note: Softening the language related to what’s permitted in invoiced billing periods is a separate conversation.

To ensure data integrity, correction records must not result in double counting of any cost- or quantity-related values. This applies regardless of the correction style or delivery mechanism used.

All correction charge records adhere to the correction handling requirements listed below.

## Attribute ID

CorrectionHandling

## Attribute Name

Correction Handling

## Description

Defines how updates to previously provided charge records are represented in FOCUS datasets.

## Requirements

* Invoice MUST be considered finalized and immutable once issued.
* Once the associated invoice is issued, each underlying charge record adheres to the following additional requirements:
  * Charge record MUST be considered finalized and immutable.
  * Charge record MUST NOT be updated, deleted, or omitted.
* Billing period MUST be considered invoiced and closed once all invoices for that period are issued.
* Additional charge records MUST NOT be associated with an invoice once it is issued.
* Additional charge records MUST NOT be associated with a billing period once it is invoiced and closed.

> 

Reviewer: These spaces make the entire bullet-point list get vertically spaced.  You may want to add subheaders if you want to delineate between sections -- otherwise the formatting gets weird.

Author: added subheaders. resolving.

# Correction Handling

## Overview

### Definition and Scope of Corrections

> Correction Handling attribute defines how updates to previously provided charge records are represented in FOCUS datasets.

Reviewer: Per our discussion in Sep 30 TF1: this is not necessarily accurate, per the point you made, @ijurica, around late-arriving charges.  I think this needs to become something like:

```suggestion
Correction Handling attribute defines how updates to invoiced charges and/or closed billing periods are represented in FOCUS datasets.
```

Author: Although the previous discussion in TF-1 was focused on corrections to charges in previously *closed billing periods* and corrections to *invoiced charges* (charges associated with issued invoices), the Correction Handling attribute should address **corrections to all previously provided charges**, including those in the current billing period. 

For instance, we are describing that the append-only mechanism does not allow updating any previously provided charge, whereas the replacement mechanism does allow such updates. Therefore, I’d suggest we keep the current phrasing.

## Attribute ID

CorrectionHandling

## Attribute Name

Correction Handling

## Description

Defines how updates to previously provided [*charges*](#glossary:charge) are represented in FOCUS datasets.

## Requirements

All corrections adhere to the following requirements:

* Correction handling implementation MUST support auditability by enabling traceability from the original record through all subsequent corrections for invoiced billing periods.
* Correction handling implementation MUST ensure that the delivery mechanisms and correction styles in use are documented within the Data Generator documentation.
* ChargeClass MUST be null when the *charge* does not represent a correction to a previously invoiced *billing period*.
* Correction MUST NOT result in double counting of any cost- or quantity-related values.
* Corrections to charges from a previously invoiced and closed billing period adhere to the following additional requirements:
  * ChargeClass MUST be "Correction".
  * Correction MUST NOT replace or omit the original record.

>   * Corrected row(s) MUST be assigned to a different InvoiceId than the original record.

Author: ```suggestion
  * Corrected row(s) MUST be assigned to a different InvoiceId than the original record, unless a replacement approach is explicitly requested by the end-user.
```
